### ISSUE TITLE
perf(training): replace O(n) replay eviction with bounded deque in distributional RL trainer

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#6508]
Parent issue: #6482 (finding #3). Sibling #6488 owns finding #1 (predictive-planner DataLoader); the AMP finding #2 is a separate child.

Replace the O(n) `replay.pop(0)` eviction in the QR-DQN custom training loop with `collections.deque(maxlen=config.dqn.replay_size)` for O(1) amortized eviction, preserving FIFO order and `random.sample`-based minibatch sampling.

```yaml
archetype: implementation
evidence_tier: smoke
linked_policy:
  - docs/dev_guide.md
```

## Scope
- In scripts/training/train_distributional_rl.py, change the `replay` container from `list` to `collections.deque(maxlen=config.dqn.replay_size)` and remove the now-redundant manual `if len(replay) > config.dqn.replay_size: replay.pop(0)` check.
- Update the `replay` type annotations (e.g. `_sample_replay` at ~lines 332-333 and the main loop at ~line 423) to reflect deque/Sequence.

## Non-goals
- No AMP/mixed-precision change (separate child).
- No predictive-planner DataLoader change (owned by #6488).
- No change to replay_size semantics, sampling distribution, loss math, gradient steps, or checkpoint/manifest output.

## Definition of done
- [ ] Replay buffer caps at config.dqn.replay_size with FIFO eviction via deque maxlen.
- [ ] random.sample minibatch sampling unchanged in distribution.
- [ ] CPU dry-run path remains deterministic and unchanged.
- [ ] Focused test added; ruff check/format pass.

## Factory v2 implementation contract
```json
{
  "allowed_paths": [
    "scripts/training/train_distributional_rl.py",
    "tests/training"
  ],
  "forbidden_paths": [
    "scripts/training/train_predictive_planner.py",
    "robot_sf/training/diffusion_policy.py",
    "robot_sf/training/oracle_imitation_bc_smoke.py",
    "robot_sf/benchmark",
    "scripts/benchmark",
    "configs/benchmark"
  ],
  "objective": "In scripts/training/train_distributional_rl.py, replace the `replay` list and its manual `if len(replay) > config.dqn.replay_size: replay.pop(0)` eviction (currently lines ~423 and 440-441) with collections.deque(maxlen=config.dqn.replay_size), dropping the redundant size check so eviction is O(1) amortized and FIFO. Update the `replay` type annotations in _sample_replay (~lines 332-333) and the main loop (~line 423) to reflect deque or Sequence. Minibatch sampling via random.sample(replay, k=batch_size) must remain valid and unchanged in distribution.",
  "stop_conditions": [
    "No change to replay_size semantics, sampling distribution, batch composition, loss/quantile_huber math, gradient steps, target-network update cadence, checkpoint/manifest output, or config schema.",
    "FIFO eviction order is preserved: deque(maxlen=...) must evict the oldest appended transition, matching the prior pop(0) behavior.",
    "CPU dry-run path remains deterministic and unchanged; no GPU hardware is required to validate.",
    "Do not touch the predictive-planner DataLoader path (owned by #6488) or add any AMP/autocast code (separate child).",
    "This child edits scripts/training/train_distributional_rl.py only in the replay data-structure region; it is file-independent of the AMP child (which edits train_predictive_planner.py) and may run as a parallel lane."
  ],
  "validation": [
    "Add or extend a focused test under tests/training asserting the replay buffer caps at config.dqn.replay_size, evicts FIFO (oldest appended transition first), and that random.sample still draws batch_size items from the capped deque.",
    "uv run pytest tests/training/test_train_distributional_rl_dry_run.py tests/training/test_train_distributional_rl_config.py -v",
    "uv run ruff check scripts/training/train_distributional_rl.py tests/training",
    "uv run ruff format --check scripts/training/train_distributional_rl.py tests/training",
    "git diff --check"
  ]
}
```

**Comment by ll7** [comment]
Draft PR opened for coordinator verification: https://github.com/ll7/robot_sf_ll7/pull/6524 (branch `orchestrator/ll7-lease-6508-ae525c215987`, head `fb4e83df`).

**Status: complete (draft).** Implements the bounded-deque replay refactor:
- `replay` is now `collections.deque(maxlen=config.dqn.replay_size)`; the manual `if len(replay) > config.dqn.replay_size: replay.pop(0)` check is removed. Eviction is O(1) amortized and FIFO (deque `maxlen` evicts oldest-first, matching the prior `pop(0)`).
- `_sample_replay` widened `list[...]` → `Sequence[...]`; main-loop `replay` annotated `deque[...]`. `random.sample(replay, k=batch_size)` is unchanged in distribution.

**Invariants preserved:** replay_size semantics, sampling distribution, loss/quantile-huber math, gradient steps, target-network cadence, checkpoint/manifest output, and config schema are untouched. Seeded trajectory is output-identical before vs after (`final_loss`/`train_steps` match exactly on the CPU path). No AMP code; predictive-planner DataLoader path untouched.

**Validation (all pass):**
- `uv run pytest tests/training/test_train_distributional_rl_dry_run.py tests/training/test_train_distributional_rl_config.py -v` → 7 passed
- New `tests/training/test_train_distributional_rl_replay_buffer.py` → 4 passed (caps at replay_size, FIFO eviction, `random.sample` on capped deque, deterministic with eviction active)
- `uv run ruff check` / `uv run ruff format --check` / `git diff --check` → clean
- Perf contract satisfied: eviction op ~13x faster/eviction at the default replay_size=4096 (O(n)→O(1)); end-to-end on this gradient-bound diagnostic loop is unchanged at smoke scale.

External artifact bundle: `/home/luttkule/.local/state/ll7-factory/runs/ll7-lease-6508-ae525c215987/` (`result.json`, `validation.json`, `RESULT.md`, `diffstat.txt`). Left as draft per lease for independent verification and promotion.
